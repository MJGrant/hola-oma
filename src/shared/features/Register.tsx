import React, { useState, useContext } from "react";
import Avatar from '@material-ui/core/Avatar';
import Button from '@material-ui/core/Button';
import CssBaseline from '@material-ui/core/CssBaseline';

import Link from '@material-ui/core/Link';
import Grid from '@material-ui/core/Grid';
import Box from '@material-ui/core/Box';
import PersonIcon from '@material-ui/icons/Person';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import Container from '@material-ui/core/Container';

import Copyright from 'shared/components/Copyright';

import { AuthContext } from "../../App";

import 'firebase/auth'; // for authentication
import 'firebase/firestore'; // if database type is firestore, import this 
import 'firebase/database'; // for additional user properties, like role 

import { RouteComponentProps } from 'react-router-dom'; // give us 'history' object

import { createNewUserWithEmailAndPassword, createNewUserWithGoogleCredentials } from "services/user";
import BigInput from "shared/components/BigInput/BigInput";

interface IRegister extends RouteComponentProps<any> {
  // empty for now 
  // got help here: https://stackoverflow.com/questions/49342390/typescript-how-to-add-type-check-for-history-object-in-react 
}

const Register: React.FC<IRegister> = ({ history }) => {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [error, setErrors] = useState("");

  const Auth = useContext(AuthContext);

  const updateEmail = (e: any) => {
    setEmail(e.target.value);
  }

  const updatePassword = (e: any) => {
    setPassword(e.target.value);
  }

  /* EMAIL & PASSWORD ACCOUNT CREATION */
  const handleForm = async (e: any) => {
    e.preventDefault();

    try {
      const userCreated = await createNewUserWithEmailAndPassword(email, password);
      
      if (userCreated) {
        Auth?.setLoggedIn(true);
        if (history) history.push('/registerDetails');
      }
    } catch(e) {
      setErrors(e.message);
    }
  }

  /* JOIN USING GOOGLE ACCOUNT */
  const handleGoogleJoin = async () => {
    try {
      const userCreated = await createNewUserWithGoogleCredentials();
      if (userCreated) {
        console.log("Created user, going to registerDetails: ", userCreated);
        Auth?.setLoggedIn(true);
        if (history) history.push('/registerDetails');
      } 
    } catch(e) {
      setErrors(e.message);
    }
  }

  const useStyles = makeStyles((theme) => ({
    paper: {
      marginTop: theme.spacing(8),
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'center',
    },
    avatar: {
      margin: theme.spacing(1),
      backgroundColor: theme.palette.secondary.main,
    },
    form: {
      width: '100%', // Fix IE 11 issue.
      marginTop: theme.spacing(3),
    },
    submit: {
      margin: theme.spacing(3, 0, 2),
    },
  }));

  const classes = useStyles();

  return (
    <Container component="main" maxWidth="xs">
      <CssBaseline />
      <div className={classes.paper}>
        <Avatar className={classes.avatar}>
          <PersonIcon />
        </Avatar>
        <Typography component="h1" variant="h5">
          Register a new account
        </Typography>
        {/* Switch to "Sign In" page */ }

        <Grid container justify="center">
          <Grid item>
            <Link href="/login" variant="body2">
              Already have an account? Log in instead
            </Link>
          </Grid>
        </Grid>

        <form onSubmit={e => handleForm(e)} className={classes.form} noValidate>

        <Grid container spacing={2}>
          
            {/* Email address */}
            <Grid item xs={12}>
              <BigInput 
                labelText="E-Mail Address"
                name="email"
                required={true} 
                value={email}
                autoFocus={true}
                autoComplete="off"
                type="text"
                onChange={updateEmail}/>

            </Grid>

            {/* Password */ }
            <Grid item xs={12}>
            <BigInput 
              labelText="Password"
              name="password"
              required={true} 
              value={password}
              autoFocus={false}
              autoComplete="off"
              type="password"
              onChange={updatePassword}/>
            </Grid>

            {/* 
            <Grid item xs={12}>
            <label>
              <input
                type="radio"
                name="accountType"
                id="receiver"
                value="receiver"
                checked={role === roles.receiver}
                onChange={e => setRole(roles.receiver)}
                />
                I want to <b>receive</b> posts
              </label>
              <br/>
              <label>
                <input
                  type="radio"
                  name="accountType"
                  id="poster"
                  value="poster"
                  onChange={e => setRole(roles.poster)}
                  />
                I want to <b>make</b> posts
              </label>
            </Grid>
              */}

          </Grid>


          <Button 
            type="submit"
            fullWidth
            variant="contained"
            color="primary"
            size="large"
            className={classes.submit}
          >
            Sign up
          </Button>

        {/* Google sign in */}
        <Grid container>
          <Grid item xs>
            <Button 
              onClick={handleGoogleJoin} 
              className="googleBtn" 
              type="button"
              variant="contained"
            >
              <img
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                alt="logo"
              />
              Join With Google
            </Button> 
          </Grid>
        </Grid>

        </form>
        <span className="error">{error}</span>
      </div>
      <Box mt={8}>
        <Copyright />      
      </Box>
    </Container>
  );
};

export default Register;