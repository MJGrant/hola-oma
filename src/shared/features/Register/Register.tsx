import React, { useState, useContext } from "react";
import Avatar from '@material-ui/core/Avatar';
import Button from '@material-ui/core/Button';

import Link from '@material-ui/core/Link';
import Grid from '@material-ui/core/Grid';
import Box from '@material-ui/core/Box';
import PersonIcon from '@material-ui/icons/Person';
import Typography from '@material-ui/core/Typography';
import Copyright from 'shared/components/Copyright';

import { AuthContext } from "../../../App";

import 'firebase/auth'; // for authentication
import 'firebase/firestore'; // if database type is firestore, import this 
import 'firebase/database'; // for additional user properties, like role 

import { createNewUserWithEmailAndPassword, createNewUserWithGoogleCredentials } from "services/user";
import BigInput from "shared/components/BigInput/BigInput";
import Alert from "@material-ui/lab/Alert";

interface IRegister {
  history?: any;
  // empty for now 
  // got help here: https://stackoverflow.com/questions/49342390/typescript-how-to-add-type-check-for-history-object-in-react 
}

const Register: React.FC<IRegister> = ({ history }) => {
  // const [displayName, setDisplayName] = useState("");
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [error, setErrors] = useState("");
  const [invalidEmail, setInvalidEmail] = useState(false);
  const [invalidPassword, setInvalidPassword] = useState(false);

  const Auth = useContext(AuthContext);

  const updateEmail = (e: any) => {
    setEmail(e.target.value);
  }

  const updatePassword = (e: any) => {
    setPassword(e.target.value);
  }

  /* EMAIL & PASSWORD ACCOUNT CREATION */
  const handleForm = async (e: any) => {
    e.preventDefault();

    if (email === "") {
      setInvalidEmail(true);
      setErrors("Please enter an email address.");
    } else if (password === "") {
      setInvalidPassword(true);
      setErrors("Please enter a password.");
    } else {
      try {
        const userCreated = await createNewUserWithEmailAndPassword(email, password);
        
        if (userCreated) {
          Auth?.setLoggedIn(true);
          if (history) history.push('/registerDetails');
        }
      } catch(e) {
        setErrors(e.message);
        setInvalidEmail(true);
        setInvalidPassword(true);
      }
    }
  }

  /* JOIN USING GOOGLE ACCOUNT */
  const handleGoogleJoin = async () => {
    try {
      const userCreated = await createNewUserWithGoogleCredentials();
      if (userCreated) {
        console.log("Created user, going to registerDetails: ", userCreated);
        Auth?.setLoggedIn(true);
        if (history) history.push('/registerDetails');
      } 
    } catch(e) {
      setErrors(e.message);
    }
  }

  return (
    <Grid container className="credentialsForm" spacing={2} justify="center">
      <Grid item xs={10} md={8}>
        <div>
          <Avatar className="formAvatar">
            <PersonIcon />
          </Avatar>
          <Typography component="h1" variant="h4">
            Register a new account
          </Typography>
          {/* Switch to "Sign In" page */ }

          <Grid container spacing={2} justify="center">
            <Grid item>
              <Link href="/login" className="bigLink">
                Already have an account? Log in instead
              </Link>
            </Grid>
          </Grid>

          <form onSubmit={e => handleForm(e)} noValidate>

          <Grid container spacing={2} justify="center">

              {/* Email address */}
              <Grid item xs={12} sm={8}>
                <BigInput
                  error={invalidEmail}
                  labelText="E-Mail Address"
                  name="email"
                  required={true} 
                  value={email}
                  autoFocus={true}
                  autoComplete="off"
                  type="email"
                  onChange={updateEmail}/>

              </Grid>

              {/* Password */ }
              <Grid item xs={12} sm={8}>
              <BigInput 
                error={invalidPassword}
                labelText="Password"
                name="password"
                required={true} 
                value={password}
                autoFocus={false}
                autoComplete="off"
                type="password"
                onChange={updatePassword}/>
              </Grid>

            <Grid item xs={12} sm={8}>
              <Button
                type="submit"
                fullWidth
                variant="contained"
                color="primary"
                size="large"
                className="bigButton"
                id="signUpButton"
              >
                Sign up
              </Button>
            </Grid>

          {error &&
            <Alert severity="error" className="errorAlert">{error}</Alert>
          }

          {/* Google sign in */}
            <Grid item xs={12} sm={8}>
              <Button 
                onClick={handleGoogleJoin} 
                className="googleBtn" 
                type="button"
                variant="contained"
              >
                <img
                  src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                  alt="logo"
                />
                Join With Google
              </Button> 
            </Grid>
            </Grid>
          </form>

        </div>
      </Grid>

      <Grid item xs={12}>
        <Box mt={6}>
          <Copyright />      
        </Box>
      </Grid>

    </Grid>
  );
};

export default Register;