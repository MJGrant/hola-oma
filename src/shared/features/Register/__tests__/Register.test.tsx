import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';

// the component to test 
import Register from '../Register';

test('The registration page is properly formed', async() => {
  // arrange
  // container = array of DOM nodes on the page
  // getByText = built-in utility function for getting elements by the text they contain 
  // getByLabelText = not used below, but another option for getting form fields by label 
  const { container, getByText } = 
    render(
      <Register />
  )

  //expect(container).toMatchSnapshot();
  
  const header = getByText('Register a new account');
  const bigInputLabels = container.querySelectorAll('.bigInputLabel'); // should be 2 of them 
  const signUpButton = getByText('Sign up');

  const errorAlert = container.querySelectorAll('.errorAlert'); // get by ID 
  
  // act
  // nothing to click or "do" here, moving along to asserts...

  // assert
  expect(header).toBeInTheDocument();
  expect(bigInputLabels.length).toBe(2);
  expect(bigInputLabels[0]).toHaveTextContent("E-Mail Address");
  expect(bigInputLabels[1]).toHaveTextContent("Password");

  expect(signUpButton).toBeInTheDocument();

  expect(errorAlert.length).toBe(0); // error shouldn't show by default 
});

// todo: user tries to submit empty form
test('The registration form cannot be submitted empty', async() => {
  // arrange
  const { container, getByText } = 
    render(
      <Register />
  )
  const signUpButton = getByText('Sign up');

  // act 
  // click the register button 
  fireEvent.click(signUpButton);

  await waitFor(() => container.querySelectorAll('.errorAlert'));

  // assert
  expect(container.querySelectorAll('.errorAlert').length).toBe(1); // error alert should be visible now

});

// todo: user tries to submit form with no password
// todo: mock server response if user tries to register an email address that exists already