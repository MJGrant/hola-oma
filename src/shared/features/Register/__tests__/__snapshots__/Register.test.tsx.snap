// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The registration page is properly formed 1`] = `<div />`;
